### PR TITLE
Jira/dev 40 derive address instruction

### PR DESCRIPTION
**Pull request by Kazopl**
## Summary by CodeRabbit

- **New Features**
	- Introduced a data model for parsed address parameters, supporting Byron and Shelley address types.
	- Added a data source abstraction for spending and staking data, enhancing address derivation capabilities.
	- Implemented new enums for spending data source types, staking data source types, and address types to improve data handling and serialization.
- **Enhancements**
	- Updated address derivation logic to utilize parsed address parameters, streamlining the process.
	- Enhanced native script hash operation with improved serialization logic for slot values.
- **Refactor**
	- Changed certain variables from `const` to `final` for better flexibility.
	- Refactored classes to use `BigInt` for slot values, accommodating larger numbers and negative values.
	- Consolidated serialization logic into utility methods for spending and staking data sources.
- **Bug Fixes**
	- Implemented validation checks for slot parameters in native scripts to prevent invalid data entries.

### CODE REVIEW

File: lib/src/models/staking_data_source.dart [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    required String keyHashHex,
+  }) = StakingDataSourceKeyHash;
+  const factory StakingDataSource.blockchainPointer({
+    required int blockIndex,
+    required int txIndex,
+    required int certificateIndex,
+  }) = StakingDataSourceBlockchainPointer;
+  const factory StakingDataSource.scriptHash({
+    required String scriptHashHex,
+  }) = StakingDataSourceScriptHash;
```
Comment (AlexDochioiu) [inline lib/src/models/staking_data_source.dart:23]: An empty line in between constructors / factories / methods makes everything a lot more readable.

File: lib/src/models/spending_data_source.dart [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+@freezed
+sealed class SpendingDataSource with _$SpendingDataSource {
+  const SpendingDataSource._();
+  const factory SpendingDataSource.none() = SpendingDataSourceNone;
+  const factory SpendingDataSource.path({
+    required List<int> path,
+  }) = SpendingDataSourcePath;
+  const factory SpendingDataSource.scriptHash({
+    required String scriptHashHex,
+  }) = SpendingDataSourceScriptHash;
```
Comment (AlexDochioiu) [inline lib/src/models/spending_data_source.dart:14]: An empty line in between constructors / factories / methods makes everything a lot more readable.

File: lib/src/cardano_ledger_application.dart [hunk 3]
```diff
@@ -128,32 +129,37 @@ class CardanoLedgerApp {
...
-    // Derivation path for shelley accounts
-    final List<int> bip32PaymentPath = [
+  Future<String> deriveAddress(LedgerDevice device, {bool displayOnDevice = false}) async {
+    final bip32PaymentPath = [
       harden + 1852,
       harden + 1815,
       harden + accountIndex,
       0,
-      0,
+      1,
```
[the remark this answers is not in the document]
Comment (Kazopl) [reply lib/src/cardano_ledger_application.dart:138]: yes

File: lib/src/models/spending_data_source.dart (hunk 2 above)
Comment (Kazopl) [reply lib/src/models/spending_data_source.dart]: done

File: lib/src/models/staking_data_source.dart (hunk 1 above)
Comment (Kazopl) [reply lib/src/models/staking_data_source.dart]: done

**Pull request merged**